File: liang_barsky_algorithm/line_clipping.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Universidad Nacional Autonoma de Mexico / Facultad de Ingenieria
# Computación Gráfica
# Alumnos:
# Barriga Martínez Diego Alberto
# Vilchis Oropeza Luis Alberto
# Hecho en Python 3.5


class Point:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def to_string(self):
        return "({0}, {1})".format(self.x, self.y)


class Window:
    def __init__(self, point, width=0, height=0):
        self.point = point
        self.width = width
        self.height = height

    def get_x_left(self):
        return self.point.get_x()

    def get_x_right(self):
        return self.point.get_x() + self.width

    def get_y_bottom(self):
        return self.point.get_y()

    def get_y_top(self):
        return self.point.get_y() + self.height

    def is_inside_window(self, point):
        if point.get_x() < self.get_x_left() or point.get_x() > self.get_x_right():
            return False
        elif point.get_y() < self.get_y_bottom() or point.get_y() > self.get_y_top():
            return False
        else:
            return True


class Line:
    def __init__(self, point_1, point_2):
        self.point_1 = point_1
        self.point_2 = point_2

    def get_x_i(self):
        return self.point_1.get_x()

    def get_x_f(self):
        return self.point_2.get_x()

    def get_y_i(self):
        return self.point_1.get_y()

    def get_y_f(self):
        return self.point_2.get_y()

    def is_vertical(self):
        return True if (self.get_x_f() - self.get_x_i()) == 0 else False

    def is_horizontal(self):
        return True if (self.get_y_f() - self.get_y_i()) == 0 else False

    def is_parallel_to_axis(self):
        return True if self.is_vertical() or self.is_horizontal() else False


def get_data():
    points = list()
    with open('input.txt', 'r') as file:
        for line in file:
            data = list(map(int, line.rstrip("\n").split(" ")))
            points.append(Point(data[0], data[1]))
    return points


def get_lines(points):
    lines = []
    for (p1, p2) in zip(points[::2], points[1::2]):
        lines.append(Line(p1, p2))
    return lines


def left_analysis(line, w):
    if line.is_parallel_to_axis():
        return None
    u = (w.get_x_left() - line.get_x_i())/(line.get_x_f() - line.get_x_i())
    if u < 0 or u > 1:
        return None
    y = line.get_y_i() + u*(line.get_y_f() - line.get_y_i())
    clipping_point = Point(w.get_x_left(), y)
    if not w.is_inside_window(clipping_point):
        return None
    return clipping_point


def right_analysis(line, w):
    if line.is_parallel_to_axis():
        return None
    u = (w.get_x_right() - line.get_x_i())/(line.get_x_f() - line.get_x_i())
    if u < 0 or u > 1:
        return None
    y = line.get_y_i() + u*(line.get_y_f() - line.get_y_i())
    clipping_point = Point(w.get_x_right(), y)
    if not w.is_inside_window(clipping_point):
        return None
    return clipping_point


def bottom_analysis(line, w):
    if line.is_parallel_to_axis():
        return None
    u = (w.get_y_bottom() - line.get_y_i())/(line.get_y_f() - line.get_y_i())
    if u < 0 or u > 1:
        return None
    x = line.get_x_i() + u*(line.get_x_f() - line.get_x_i())
    clipping_point = Point(x, w.get_y_bottom())
    if not w.is_inside_window(clipping_point):
        return None
    return clipping_point


def top_analysis(line, w):
    if line.is_parallel_to_axis():
        return None
    u = (w.get_y_top() - line.get_y_i())/(line.get_y_f() - line.get_y_i())
    if u < 0 or u > 1:
        return None
    x = line.get_x_i() + u*(line.get_x_f() - line.get_x_i())
    clipping_point = Point(x, w.get_y_top())
    if not w.is_inside_window(clipping_point):
        return None
    return clipping_point


def liang_barsky(line, window):
    intersections = dict()
    intersections["left"] = left_analysis(line, window)
    intersections["right"] = right_analysis(line, window)
    intersections["top"] = top_analysis(line, window)
    intersections["bottom"] = bottom_analysis(line, window)
    return intersections


def line_clipper(lines, window):
    line_counter = 1
    for l in lines:
        clipping_points = liang_barsky(l, window)
        store_information(clipping_points["left"], "Left", line_counter)
        store_information(clipping_points["right"], "Right", line_counter)
        store_information(clipping_points["top"], "Top", line_counter)
        store_information(clipping_points["bottom"], "Bottom", line_counter)
        line_counter += 1


def store_information(clipping_point, side, line_id):
    with open('output.txt', 'a') as file:
        message = "{0} limit, line {1}: {2}\n"
        if clipping_point:
            message = message.format(side, line_id, clipping_point.to_string())
        else:
            message = message.format(side, line_id, "it has no clipping point.")
        file.write(message)


def main():
    lines = get_lines(get_data())
    print("===Enter Window data===")
    w_point = list(map(int, input("Point (separate by space): ").split(" ")))
    width = int(input("Width: "))
    height = int(input("Height: "))
    w = Window(Point(w_point[0], w_point[1]), width, height)
    line_clipper(lines, w)


if __name__ == '__main__':
    main()

File: liang_barsky_algorithm/line_clipping.c
#include <stdio.h>
#include <stdlib.h>

#ifndef LINE_CLIPPING_DATA
struct _point {
	float x;
	float y;
};
typedef struct _point POINT;

struct _line {
	POINT start;
	POINT end;
};
typedef struct _line LINE;

struct _window {
	POINT point;
	int width;
	int height;
};
typedef struct _window WINDOW;
#endif

#ifndef LINE_CLIPPING_DATA_OPERATIONS

#define window_x_left(w) w.point.x
#define window_x_right(w) (w.point.x+w.width)
#define window_y_bottom(w) w.point.y
#define window_y_top(w) (w.point.y+w.height)

#define line_x_i(l) l.start.x
#define line_x_f(l) l.end.x
#define line_y_i(l) l.start.y
#define line_y_f(l) l.end.y

#define line_x_dif(l) (l.end.x-l.start.x)
#define line_y_dif(l) (l.end.y-l.start.y)

POINT* left_analysis(LINE, WINDOW);
POINT* right_analysis(LINE, WINDOW);
POINT* top_analysis(LINE, WINDOW);
POINT* bottom_analysis(LINE, WINDOW);
void liang_barsky(FILE*, FILE*, char**);
int is_inside_window(POINT*, WINDOW);
#endif



POINT* left_analysis(LINE l, WINDOW w) {
	float u, y;
	POINT *pclipping;
	if (line_x_dif(l) == 0) {
		return NULL;
	}
	u = (float)((window_x_left(w) - line_x_i(l)) / line_x_dif(l));
	if (u < 0 || u > 1) {
		return NULL;
	}
	y = line_y_i(l) + u*(line_y_dif(l));
	pclipping = (POINT *) malloc(sizeof(POINT));
	pclipping->x = window_x_left(w);
	pclipping->y = y;
	if (!is_inside_window(pclipping,w)) {
		return NULL;
	}
	return pclipping;
}

POINT* right_analysis(LINE l, WINDOW w) {
	float u, y;
	POINT *pclipping;
	if (line_x_dif(l) == 0) {
		return NULL;
	}
	u = (float)((window_x_right(w) - line_x_i(l)) / line_x_dif(l));
	if (u < 0 || u > 1) {
		return NULL;
	}
	y = line_y_i(l) + u*(line_y_dif(l));
	pclipping = (POINT *) malloc(sizeof(POINT));
	pclipping->x = window_x_right(w);
	pclipping->y = y;
	if (!is_inside_window(pclipping,w)) {
		return NULL;
	}
	return pclipping;
}

POINT* bottom_analysis(LINE l, WINDOW w) {
	float u, x;
	POINT *pclipping;
	if (line_y_dif(l) == 0) {
		return NULL;
	}
	u = (float)((window_y_bottom(w) - line_y_i(l)) / line_y_dif(l));
	if (u < 0 || u > 1) {
		return NULL;
	}
	x = line_x_i(l) + u*(line_x_dif(l));
	pclipping = (POINT *) malloc(sizeof(POINT));
	pclipping->x = x;
	pclipping->y = window_y_bottom(w);
	if (!is_inside_window(pclipping,w)) {
		return NULL;
	}
	return pclipping;
}

POINT* top_analysis(LINE l, WINDOW w) {
	float u, x;
	POINT *pclipping;
	if (line_y_dif(l) == 0) {
		return NULL;
	}
	u = (float)((window_y_top(w) - line_y_i(l)) / line_y_dif(l));
	if (u < 0 || u > 1) {
		return NULL;
	}
	x = line_x_i(l) + u*(line_x_dif(l));
	pclipping = (POINT *) malloc(sizeof(POINT));
	pclipping->x = x;
	pclipping->y = window_y_top(w);
	if (!is_inside_window(pclipping,w)) {
		return NULL;
	}
	return pclipping;
}

void liang_barsky(FILE *input, FILE *output, char **argv) {
	int index = 0, i;
	LINE *lines = (LINE *) malloc(sizeof(LINE));
	WINDOW w;
	w.point.x = atof(argv[3]);
	w.point.y = atof(argv[4]);
	w.width = atoi(argv[5]);
	w.height = atoi(argv[6]);
	while(!feof(input)) {
		lines = (LINE*) realloc(lines, (index + 1)*sizeof(LINE));
		fscanf(input, "%f %f %f %f", &lines[index].start.x, &lines[index].start.y,
									 &lines[index].end.x, &lines[index].end.y);
		index++;
	}
	for(i = 0; i < index; i++) {
		POINT *p;
		if((p=left_analysis(lines[i],w)) != NULL)
			fprintf(output, "Left analysis: (%.3f, %.3f).\n", p->x, p->y);
		else
			fprintf(output, "Left analysis: has no clipping point.\n");
		if((p=right_analysis(lines[i],w)) != NULL)
			fprintf(output, "Right analysis: (%.3f, %.3f).\n", p->x, p->y);
		else
			fprintf(output, "Right analysis: has no clipping point.\n");
		if((p=top_analysis(lines[i],w)) != NULL)
			fprintf(output, "Top analysis: (%.3f, %.3f).\n", p->x, p->y);
		else
			fprintf(output, "Top analysis: has no clipping point.\n");
		if((p=bottom_analysis(lines[i],w)) != NULL)
			fprintf(output, "Bottom analysis: (%.3f, %.3f).\n", p->x, p->y);
		else
			fprintf(output, "Bottom analysis: has no clipping point.\n");
		fprintf(output, "\n");
	}
	free(lines);
}

int is_inside_window(POINT* p, WINDOW w) {
	if (p->x < window_x_left(w) || p->x > window_x_right(w)) {
		return 0;
	} else if (p->y < window_y_bottom(w) || p->y > window_y_top(w)) {
		return 0;
	} else {
		return 1;
	}
}

int main(int argc, char **argv) {
	FILE *input, *output;
	if (argc < 7) {
		printf("Usage: ./liang-barsky input output x y width height\n");
		exit(EXIT_FAILURE);
	}
	if (!(input = fopen(argv[1], "r"))) {
		perror("File error");
		exit(EXIT_FAILURE);
	}
	if(!(output = fopen(argv[2], "w"))) {
		perror("File error");
		exit(EXIT_FAILURE);
	}
	liang_barsky(input, output, argv);
	fclose(input);
	fclose(output);
	return 0;
}